sympy/solvers/solveset.py

```diff
@@ -120,11 +121,63 @@ def _invert_real(f, g_ys, symbol):
             return _invert_real(expo, imageset(Lambda(n, log(n)/log(base)),
                                                g_ys), symbol)
 
+    if isinstance(f, sin) or isinstance(f, csc):
+        n = Dummy('n')
+        if isinstance(g_ys, FiniteSet):
+            fi = sympify('a' + f.func.__name__)
+            # remove all the invalid elements
+            _a = []
+            for i in g_ys:
+                try:
+                    if i.is_real is S.false or \
+                        (isinstance(f, sin) and i not in Interval(-1, 1)) or \
```

Comment:
Doing `sin` and `cosec` together makes your implementation looks somewhat clumsy, consider seperating them.